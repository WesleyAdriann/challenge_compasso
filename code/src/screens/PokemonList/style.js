import styled from 'styled-components/native'

import {COLORS} from '../../style'

export const ScreenContainer = styled.SafeAreaView`
  background-color: ${COLORS.primatyGray};
  flex-grow: 1;
  border-left-width: 8px;
  border-right-width: 8px;
  border-color: ${COLORS.primaryRed};
`

export const PokemonsFlatList = styled.FlatList.attrs(() => ({
  contentContainerStyle: {
    paddingVertical: 8
  },
}))``
