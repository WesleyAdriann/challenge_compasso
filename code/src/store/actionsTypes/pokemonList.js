const prefix = 'POKEMON_LIST'
export const pokemonListTypes = {
  SETPOKEMONS: `${prefix}_SET_POKEMONS`,
  SETPAGE: `${prefix}_SET_PAGE`,
  SETPERPAGE: `${prefix}_SET_PER_PAGE`,
  HANDLECHANGETEXT: `${prefix}_HANDLE_CHANGE_TEXT`,
  SETISLOADING: `${prefix}_SET_IS_LOADING`
}
