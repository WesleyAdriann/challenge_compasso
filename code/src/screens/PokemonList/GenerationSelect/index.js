import React from 'react'
import PropTypes from 'prop-types'

import { GenerationSelectWrapper, Label, OptionsContainer, GenerationButton } from './style'

const GenerationSelect = ({ generations, selected, handleSelectGeneration }) => (
  <GenerationSelectWrapper>
    <Label>Generations</Label>
    <OptionsContainer>
      {
        generations.map((generation, index) => (
          <GenerationButton
            key={index}
            selected={selected === generation.id}
            onPress={() => handleSelectGeneration(generation.id)}
          >
            {
              selected === generation.id
                ? <GenerationButton.Icon />
                : <GenerationButton.Text>{generation.id}</GenerationButton.Text>
            }
          </GenerationButton>
        ))
      }
    </OptionsContainer>
  </GenerationSelectWrapper>
)

GenerationSelect.propTypes = {
  generations: PropTypes.arrayOf(PropTypes.object).isRequired,
  selected: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  handleSelectGeneration: PropTypes.func
}

GenerationSelect.defaultProps = {
  handleSelectGeneration: () => {},
  selected: 0
}

export default GenerationSelect
