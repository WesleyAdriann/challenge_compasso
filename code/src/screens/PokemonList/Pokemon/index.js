import React from 'react'
import PropTypes from 'prop-types'

import { string } from '../../../utils'

import { PokemonContainer, PokemonNumber, PokemonImage, PokemonName } from './style'

const Pokemon = ({ id, name, handlePushToPokemon }) => (
  <PokemonContainer onPress={handlePushToPokemon}>
    <PokemonNumber>
      #{String(id).padStart(4, '0')}
    </PokemonNumber>
    <PokemonImage
      source={{
        uri: `https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/${id}.png`
      }}
    />
    <PokemonName>{string.capitalize(name)}</PokemonName>
  </PokemonContainer>
)

Pokemon.propTypes = {
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  name: PropTypes.string.isRequired,
  handlePushToPokemon: PropTypes.func.isRequired
}

export default Pokemon
