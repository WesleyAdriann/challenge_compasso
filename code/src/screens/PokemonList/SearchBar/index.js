import React from 'react'
import PropTypes from 'prop-types'

import { IMAGES } from '../../../style'

import RoundButton from '../../../components/RoundButton'

import { SearchBarWrapper, SearchInput } from './style'

const SearchBar = ({ handleInputChange, searchValue, handleSearchPokemon, wasSearched }) => (
  <SearchBarWrapper>
    <SearchInput
      value={searchValue}
      onChangeText={handleInputChange('searchInput')}
      placeholder='Pokemon name or number'
    />
    <RoundButton onPress={handleSearchPokemon}>
      <RoundButton.Icon source={wasSearched ? IMAGES.icons.cancelIcon : IMAGES.icons.searchIcon} />
    </RoundButton>
  </SearchBarWrapper>
)

SearchBar.propTypes = {
  handleInputChange: PropTypes.func.isRequired,
  searchValue: PropTypes.string,
  handleSearchPokemon: PropTypes.func.isRequired,
  wasSearched: PropTypes.bool
}

SearchBar.defaultProps = {
  searchValue: '',
  wasSearched: false
}

export default SearchBar
