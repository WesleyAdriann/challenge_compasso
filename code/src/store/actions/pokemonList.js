import { pokemonListTypes } from '../actionsTypes'

export const setPokemons = (pokemonList) => ({
  type: pokemonListTypes.SETPOKEMONS,
  pokemonList
})

export const setGenerations = (generationList) => ({
  type: pokemonListTypes.SETGENERATIONS,
  generationList
})

export const setPage = (page) => ({
  type: pokemonListTypes.SETPAGE,
  page
})

export const handleChangeText = (text, inputName) => ({
  type: pokemonListTypes.HANDLECHANGETEXT,
  text,
  inputName
})

export const setIsLoading = (isLoading) => ({
  type: pokemonListTypes.SETISLOADING,
  isLoading
})
