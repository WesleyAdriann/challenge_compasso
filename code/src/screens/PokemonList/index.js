import React, { useEffect } from 'react'
import { Alert } from 'react-native'
import { useSelector, useDispatch } from 'react-redux'
import PropTypes from 'prop-types'

import {
  setPokemons,
  handleChangeText,
  setPage,
  setIsLoading,
  setGenerations,
  setTotalPokemons,
  setSelectedGeneration,
  setWasSearched
} from '../../store/actions/pokemonList'
import { getPokemonsList, getGenerationsList, getPokemon } from '../../services/pokeAPI'
import { ROUTES } from '../../settings'

import LoadingFull from '../../components/LoadingFull'
import Loading from '../../components/Loading'

import Pokemon from './Pokemon'
import SearchBar from './SearchBar'
import GenerationSelect from './GenerationSelect'
import { ScreenContainer, PokemonsFlatList } from './style'

let requesTimeout = 0

const PokemonList = ({ navigation }) => {
  const dispatch = useDispatch()
  const {
    pokemonList,
    page,
    perPage,
    searchInput,
    isLoading,
    generationList,
    selectedGeneration,
    totalPokemons,
    wasSearched
  } = useSelector((state) => state.pokemonList)

  /**
   * Function to request pokemon list and set in redux store
   * @param {number} pageToRequest - Page to search
   */
  const requestPokemonList = (pageToRequest) => {
    if (isNaN(pageToRequest)) return
    dispatch(setIsLoading(true))
    clearTimeout(requesTimeout)
    requesTimeout = setTimeout(() => {
      getPokemonsList(pageToRequest, perPage)
        .then(({ data }) => {
          const { results } = data
          let pokemons = results.map((poke) => {
            const id = poke.url.split('pokemon/').pop().replace(/\D/g, '')
            return { ...poke, id }
          })

          // concat pokemons
          if (pageToRequest) pokemons = [...pokemonList, ...pokemons]

          dispatch(setPokemons(pokemons))
          dispatch(setPage(pageToRequest))
          dispatch(setIsLoading(false))
          if (!totalPokemons) dispatch(setTotalPokemons(data.count))
          if (selectedGeneration) dispatch(setSelectedGeneration(0))
        })
        .catch((error) => {
          console.log('requestPokemonList error:', error)
          Alert.alert('Attention', 'There was an error loading the pokémons', [
            { text: 'Ok', onPress: () => dispatch(setIsLoading(false)) },
            { text: 'Try Again', onPress: () => requestPokemonList(pageToRequest) }
          ])
        })
    }, 500)
  }

  /**
   * Request to get list of pokemons generations
   */
  const requestPokemonGenerations = () => {
    if (generationList.length) return
    getGenerationsList()
      .then(({ data }) => {
        const { results } = data
        const generations = results.map((generation) => {
          const id = generation.url.split('generation/').pop().replace(/\D/g, '')
          return { ...generation, id }
        })
        dispatch(setGenerations(generations))
      })
      .catch((error) => {
        console.log('requestPokemonList error:', error)
        setTimeout(requestPokemonGenerations, 5000)
      })
  }

  /**
   * Request to get pokemons from one generation
   * @param {number} generationId - Id of pokemon generation
   */
  const requestPokemonsFromGeneration = (generationId) => {
    dispatch(setIsLoading(true))
    clearTimeout(requesTimeout)
    requesTimeout = setTimeout(() => {
      getGenerationsList(generationId)
        .then(({ data }) => {
          const { pokemon_species: results } = data
          const pokemons = results
            .map((poke) => {
              const id = poke.url.split('pokemon-species/').pop().replace(/\D/g, '')
              return { ...poke, id }
            })
            .sort(({ id: pokeAId }, { id: pokeBId }) => pokeAId - pokeBId)
          dispatch(setPokemons(pokemons))
          dispatch(setPage(0))
          dispatch(setSelectedGeneration(generationId))
          dispatch(setIsLoading(false))
        })
        .catch((error) => {
          console.log('requestPokemonsFromGeneration error:', error)
          Alert.alert('Attention', 'There was an error loading the pokémons', [
            { text: 'Ok', onPress: () => dispatch(setIsLoading(false)) },
            { text: 'Try Again', onPress: () => requestPokemonsFromGeneration(generationId) }
          ])
        })
    }, 500)
  }

  /**
   * Request to search pokemon using name or number
   */
  const requestPokemonSearch = () => {
    dispatch(setIsLoading(true))
    clearTimeout(requesTimeout)
    requesTimeout = setTimeout(() => {
      getPokemon(searchInput)
        .then(({ data }) => {
          const pokemon = {
            id: data.id,
            name: data.name
          }
          dispatch(setPokemons([pokemon]))
          dispatch(setPage(0))
          dispatch(setWasSearched(true))
          dispatch(setIsLoading(false))
        })
        .catch((error) => {
          console.log('requestPokemonsFromGeneration error:', error)
          if (error.response.status === 404) {
            Alert.alert('Attention', 'Pokemon not finded', [
              {
                text: 'Ok',
                onPress: () => {
                  dispatch(setIsLoading(false))
                  requestPokemonList(0)
                }
              }
            ])
          }
        })
    }, 500)
  }

  /**
   * Handler to load more pokemons in infinity scroll
   */
  const handleLoadMorePokemon = () => {
    if (selectedGeneration) return
    if (wasSearched) return
    if (page * perPage >= totalPokemons) return Alert.alert('Attention', 'All pokemons are loaded')

    const nextPageToRequest = page + 1
    requestPokemonList(nextPageToRequest)
  }

  /**
   * Handler to manipulate to get or clear pokemons from generation
   * @param {number} generationId - Id of pokemon generation
   */
  const handleSelectGeneration = (generationId) => {
    dispatch(setPokemons([]))

    if (generationId === selectedGeneration) {
      dispatch(setSelectedGeneration(0))
      requestPokemonList(0)
      return
    }

    dispatch(setWasSearched(false))
    requestPokemonsFromGeneration(generationId)
  }

  /**
   * Handler to manipulate search pokemons or clear search
   */
  const handleSearchPokemon = () => {
    if (!searchInput && !wasSearched) return Alert.alert('', 'Insert Pokemon name or number')

    dispatch(setSelectedGeneration(0))
    dispatch(setPokemons([]))
    if (wasSearched) {
      dispatch(setWasSearched(false))
      handleInputChange('searchInput')('')
      requestPokemonList(0)
      return
    }

    requestPokemonSearch()
  }

  /**
   * Handler to text input
   * @param {string} inputName - Name of input to manipulate in store
   * @param {string} text - Text from input event
   */
  const handleInputChange = (inputName) => (text) => dispatch(handleChangeText(text, inputName))

  /**
   * Handler to navigate to pokemon selected
   * @param {number | string} pokemonId - Id from pokemon
   */
  const handlePushToPokemon = (pokemonId) => () => navigation.push(ROUTES.pokemon, { pokemonId })

  useEffect(() => {
    requestPokemonList(0)
    requestPokemonGenerations()
  }, [])

  return (
    <ScreenContainer>
      <LoadingFull visible={!pokemonList.length && isLoading} />
      <SearchBar
        handleInputChange={handleInputChange}
        searchValue={searchInput}
        handleSearchPokemon={handleSearchPokemon}
        wasSearched={wasSearched}
      />
      <GenerationSelect
        generations={generationList}
        handleSelectGeneration={handleSelectGeneration}
        selected={selectedGeneration}
      />
      <PokemonsFlatList
        data={pokemonList}
        numColumns={3}
        renderItem={({ item, index }) => <Pokemon key={`${index}-${item.id}`} id={item.id} name={item.name} handlePushToPokemon={handlePushToPokemon(item.id)} /> }
        onEndReached={handleLoadMorePokemon}
        onEndReachedThreshold={0.5}
        ListFooterComponent={() => <Loading visible={!!pokemonList.length && isLoading} /> }
      />
    </ScreenContainer>
  )
}

PokemonList.propTypes = {
  navigation: PropTypes.object.isRequired
}

export default PokemonList
