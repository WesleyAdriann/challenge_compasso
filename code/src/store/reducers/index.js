import { combineReducers } from 'redux'

import pokemon from './pokemon'
import pokemonList from './pokemonList'

const reducers = combineReducers({
  pokemon,
  pokemonList
})

export default reducers
