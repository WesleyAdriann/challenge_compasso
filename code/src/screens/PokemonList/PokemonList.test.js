import React from 'react'
import renderer from 'react-test-renderer'
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux'

import { INITIAL_STATE } from '../../store/reducers/pokemonList' 

import PokemonList from './index'

describe('PokemonList screen', () => {
  
  test('renders correctly', () => {
    const mockStore = configureStore({})
    const store = mockStore({pokemonList: INITIAL_STATE})
  
    const props = {
      navigation: {
        push: jest.fn(),
      }
    }
  
    const tree = renderer.create(
      <Provider store={store}>
        <PokemonList {...props} />
      </Provider>
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

})

