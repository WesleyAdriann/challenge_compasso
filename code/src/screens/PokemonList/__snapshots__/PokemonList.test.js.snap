// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonList screen renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Array [
      Object {
        "backgroundColor": "#e7e7e6",
        "height": "100%",
      },
    ]
  }
>
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    transparent={true}
    visible={false}
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(128, 128, 128, 0.4)",
            "justifyContent": "center",
            "minHeight": 1334,
            "paddingBottom": 10,
            "paddingLeft": 10,
            "paddingRight": 10,
            "paddingTop": 10,
            "width": 750,
          },
        ]
      }
    >
      <ActivityIndicator
        animating={true}
        color="#D50000"
        hidesWhenStopped={true}
        size="large"
      />
      <Text
        style={
          Array [
            Object {
              "color": "white",
              "paddingTop": 10,
            },
          ]
        }
      >
        
      </Text>
    </View>
  </Modal>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "height": 48,
          "marginBottom": 4,
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 4,
          "paddingBottom": 4,
          "paddingLeft": 8,
          "paddingRight": 8,
          "paddingTop": 4,
        },
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Pokemon name or number"
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "backgroundColor": "white",
            "borderColor": "#050505",
            "borderRadius": 200,
            "borderWidth": 1,
            "elevation": 3,
            "flexGrow": 1,
            "paddingBottom": 8,
            "paddingLeft": 12,
            "paddingRight": 12,
            "paddingTop": 8,
          },
        ]
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#D50000",
          "borderColor": "#050505",
          "borderRadius": 200,
          "borderWidth": 1,
          "elevation": 3,
          "height": 40,
          "justifyContent": "center",
          "marginBottom": 0,
          "marginLeft": 8,
          "opacity": 1,
          "width": 40,
        }
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../src/assets/images/icons/search.png",
          }
        }
        style={
          Array [
            Object {
              "height": 16,
              "resizeMode": "cover",
              "width": 16,
            },
          ]
        }
      />
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "marginBottom": 8,
          "marginLeft": 8,
          "marginRight": 8,
          "marginTop": 0,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#050505",
            "fontFamily": "Roboto",
            "fontSize": 12,
            "marginBottom": 2,
            "textAlign": "center",
          },
        ]
      }
    >
      Generations
    </Text>
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
            "flexWrap": "wrap",
            "justifyContent": "space-between",
            "paddingRight": 8,
          },
        ]
      }
    />
  </View>
  <RCTScrollView
    ListFooterComponent={[Function]}
    contentContainerStyle={
      Object {
        "paddingBottom": 8,
      }
    }
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0.5}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Array [
        Object {
          "flexGrow": 1,
        },
      ]
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
      />
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
